math/src/main/java/smile/math/matrix/SparseMatrix.java

```diff
@@ -175,11 +180,204 @@ public int size() {
+    public int[] getColIndex() {
+        return colIndex;
+    }
+
+    /**
+     * Calls a lambda with each non-zero value. This will be a bit faster than using a stream
+     * because we can avoid boxing up the coordinates of the element being processed, but it
+     * will be considerably less general.
+     * <p>
+     * Note that the action could be called on values that are either effectively or actually
+     * zero. The only guarantee is that no values that are known to be zero based on the
+     * structure of the matrix will be processed.
+     *
+     * @param action Action to perform on each non-zero, typically this is a lambda.
+     */
+    @SuppressWarnings("WeakerAccess")
+    public void foreachNonzero(MatrixElementConsumer action) {
+        for (int j = 0; j < ncols; j++) {
+            for (int k = colIndex[j]; k < colIndex[j + 1]; k++) {
+                int i = rowIndex[k];
+                action.eval(i, j, x[k]);
+            }
+        }
+    }
+
+
+    /**
+     * Calls a lambda with each non-zero value. This will be a bit faster than using a stream
+     * because we can avoid boxing up the coordinates of the element being processed, but it
+     * will be considerably less general.
+     * <p>
+     * Note that the action could be called on values that are either effectively or actually
+     * zero. The only guarantee is that no values that are known to be zero based on the
+     * structure of the matrix will be processed.
+     *
+     * @param startColumn The first column to scan.
+     * @param endColumn   One past the last column to scan.
+     * @param action      Action to perform on each non-zero, typically this is a lambda.
+     */
+    @SuppressWarnings("WeakerAccess")
+    public void foreachNonzero(int startColumn, int endColumn, MatrixElementConsumer action) {
+        if (startColumn < 0 || startColumn >= ncols) {
+            throw new IllegalArgumentException("Start column must be in range [0,ncols)");
+        }
+        if (endColumn < 0 || endColumn > ncols) {
+            throw new IllegalArgumentException("End column must be in range [startColumn,ncols]");
+        }
+
+        for (int j = startColumn; j < endColumn; j++) {
+            for (int k = colIndex[j]; k < colIndex[j + 1]; k++) {
+                int i = rowIndex[k];
+                action.eval(i, j, x[k]);
+            }
+        }
+    }
+
+    /**
+     * Provides a stream over all of the non-zero elements of a sparse matrix.
+     */
+    @SuppressWarnings("WeakerAccess")
+    public Stream<SparseMatrixEntry> nonzeros() {
+        return StreamSupport.stream(new SparseMatrixSpliterator(this, 0, ncols), false);
+    }
+
+    /**
+     * Provides a stream over all of the non-zero elements of range of columns of a sparse matrix.
+     *
+     * @param startColumn The first column to scan
+     * @param endColumn   The first column after the ones that we should scan.
+     */
+    @SuppressWarnings("WeakerAccess")
```

Comment:
I am really concerned that there are so many SuppressWarnings, which indicates some design flaws.

Reply:
It just indicates that the only user of these functions is a test case that is in the same package while the ultimate user of the function is in user code.

Reply:
Can you elaborate?

Reply:
A method like foreachNonZero needs to be public because it will be used by applications. That is the way that the code currently is.

But IntelliJ is analyzing just the Smile code and it notices that the method is only used in the SparseMatrixTest class and further notices that this test case is in the same package as the original code. It therefore incorrectly flags the definition of the function as having broader access rights than are necessary. Thus, the SuppressWarnings.

I could suppress the warning across the entire project for my own use, but that would leave the warnings visible to others who work on the code using similar toolsets. I think it is better to leave the warning levels at the default level so I see what everybody else will see. At that point it is good practice to mark some warnings as spurious so that any real warnings will be visible.